community/index/src/test/java/org/neo4j/index/gbptree/GBPTreeIT.java

```diff
@@ -498,6 +498,196 @@ else if ( lastSeenKey > nextToSee )
+
+        // Thread communication
+        AtomicInteger currentWriteIteration = new AtomicInteger( 0 );
+        AtomicLong lastWrittenKey = new AtomicLong( -1 );
+        CountDownLatch readerReadySignal = new CountDownLatch( readers );
+        CountDownLatch startSignal = new CountDownLatch( 1 );
+        AtomicBoolean endSignal = new AtomicBoolean();
+        AtomicBoolean failHalt = new AtomicBoolean(); // Readers signal to writer that there is a failure
+        AtomicReference<Throwable> readerError = new AtomicReference<>();
+        AtomicInteger numberOfReads = new AtomicInteger();
+
+        // given
+        index = createIndex( 256 );
+
+        Runnable reader = () -> {
+            int numberOfLocalReads = 0;
+            try
+            {
+                readerReadySignal.countDown();
+                while ( currentWriteIteration.get() < 1 )
+                {
+                    startSignal.await( 5, SECONDS );
+                }
+
+                while ( !endSignal.get() && !failHalt.get() )
+                {
+                    // Read one go, we should see up to highId
+
+                    // Kept for SeekCursor life
+                    int iterationExpectedToSee = currentWriteIteration.get() - 1;
+                    int rangeModulus = iterationExpectedToSee % nbrOfGroups;
+                    long start = maxRange( nbrOfGroups, rangeWidth, iterationExpectedToSee );
+                    long end = minRange( nbrOfGroups, rangeWidth, iterationExpectedToSee );
+
+                    // Updated during SeekCursor life
+                    long lastSeenKey = -1;
+                    long nextToSeeBase = start;
+                    long nextToSeeDelta = 0;
+                    long nextToSee = nextToSeeBase - nextToSeeDelta;
+                    long lastWrittenBeforeStart;
+                    long lastWrittenWhenFinished;
+                    long lastWrittenBeforeTraversingTree;
+
+                    lastWrittenBeforeTraversingTree = lastWrittenKey.get();
+                    try ( RawCursor<Hit<MutableLong,MutableLong>,IOException> cursor =
+                                  index.seek( new MutableLong( start ), new MutableLong( end ) ) )
+                    {
+                        lastWrittenWhenFinished = lastWrittenKey.get();
+                        while ( cursor.next() )
+                        {
+                            lastWrittenBeforeStart = lastWrittenWhenFinished;
+                            lastWrittenWhenFinished = lastWrittenKey.get();
+
+                            lastSeenKey = cursor.get().key().longValue();
+                            long lastSeenValue = cursor.get().value().longValue();
+                            if ( lastSeenKey != lastSeenValue )
+                            {
+                                failHalt.set( true );
+                                fail( String.format( "Read mismatching key value pair, key=%d, value=%d%n",
```

Comment:
failing in separate thread will not gonna fail original test, will it? Maybe save message somewhere and fail main test thread instead?